### ISSUE TITLE
Multi GPU support on CustomObject Training

### ISSUE THREAD

**Issue by rola93** [OlafenwaMoses/ImageAI#288]
This was asked [in this thread](https://github.com/OlafenwaMoses/ImageAI/issues/9#issuecomment-461965526) but wasn't answered.

I'ld like to know if this supports multiple gpu or not, and how.

Reading the code, I come across with [this line](https://github.com/OlafenwaMoses/ImageAI/blob/a8fe7bbac6777c23dce6e1cf31d3ebd8ff4584ca/imageai/Detection/Custom/__init__.py#L56):

``` python
class DetectionModelTrainer:

    """
    This is the Detection Model training class, which allows you to train object detection models
    on image datasets that are in Pascal VOC annotation format, using the YOLOv3.
    """

    def __init__(self):
        # define other parameters here....
        self.__train_gpus = "0,1"
```
So I have some questions with respect to this:

1) If I only do have one GPU, wouldn't it be better to say it explicitely? changing `self.__train_gpus = "0,1"` to `self.__train_gpus = "0"`. I know that, according to the docs, if no GPU:1 is provided, TF will assign its work to GPU:0... and if GPU:0 is not available, assign the work to CPU:0, which is always provided, rigth? But, if I do know before hand that only one GPU is provided, isn't better & more performant to assign all work to this GPU?

2) If I have more than 2 GPUs, the rest of them isn't going to be loaded, rigth?

So, my final guess/question: is it better to explicitly say the number of GPUs? To use more than 2, all you need is to re define this line: `self.__train_gpus = "0,1"`??

Thanks in advance

**Comment by rola93** [comment]
Well, I keep reading the code, and I did the following change: `self.__train_gpus = "0,1"` to `self.__train_gpus = "0,1,2,3,4,5,6,7"`

Then, as the batch's workload is distributed among each available GPU, increased its batch_size to 32 (and therefore, 4 samples per GPU).

Then I run the well-known `watch -n 0.5 nvidia-smi` command, and this is its output:

![GPU usage](https://i.imgur.com/PpN8AEe.png)

All eigth GPUs are being used.

Y also tried to force GPU usage to just 4, with `self.__train_gpus = "0,1,2,3"` and what I see is the following:

![GPU usage](https://i.imgur.com/Ra1Q73i.png)

Notice that it is actually using 5 GPUs instead of four. However, I'm pretty sure that this additional GPU is just an independent issue: even when no work is being done on GPU, this GPU seems to be working:

![no usage GPU](https://i.imgur.com/RpCgukr.png)

Something similar happens when I used only one GPU: it looks like doing work, even whe it's not.

To sum up:

There should be some way to specify which GPU to use while training. Will write a PR later

**Comment by rola93** [comment]
Well, PR is [here](https://github.com/OlafenwaMoses/ImageAI/pull/291) so will close this issue

**Comment by OlafenwaMoses** [comment]
Thanks very much for this breakdown work on the issue. I really appreciate the analysis you made on this. I have merged the PR and looking forward to more of such.
